src/network/networkd-dhcp6.c

```diff
@@ -35,6 +36,93 @@ static bool dhcp6_get_prefix_delegation(Link *link) {
                       RADV_PREFIX_DELEGATION_BOTH);
 }
 
+static bool dhcp6_has_preferred_subnet_id(Link *link) {
+        assert(link->network);
+        return link->network->router_prefix_subnet_id > -1;
+}
+
+static int dhcp6_get_preferred_delegated_prefix(
+                Manager* manager,
+                Link *link,
+                const struct in6_addr *pd_prefix,
+                uint8_t pd_prefix_len,
+                struct in6_addr *ret_addr) {
+        int r;
+        union in_addr_union prefix;
+        union in_addr_union pd_prefix_union;
+        const uint8_t prefix_bits = 64 - pd_prefix_len;
+        const uint64_t n_prefixes = UINT64_C(1) << prefix_bits;
+        const int64_t subnet_id = link->network->router_prefix_subnet_id;
+        _cleanup_free_ char *assigned_buf = NULL;
+
+        pd_prefix_union.in6 = *pd_prefix;
+
+        /* We start off with the original PD prefix we have been assigned and
+         * iterate from there */
+        prefix.in6 = *pd_prefix;
+        log_link_debug(link, "subnet id %" PRIi64, subnet_id);
+
+        if (subnet_id > -1) {
+                /* If the link has a preference for a particular subnet id try to allocate that */
+                if ((uint64_t)subnet_id >= n_prefixes) {
+                        return log_link_debug_errno(link,
+                                        SYNTHETIC_ERRNO(ERANGE),
+                                        "subnet id %" PRIi64 " is out of range. Only have %" PRIu64 " subnets",
+                                         subnet_id,
+                                         n_prefixes);
+                } else {
```

Comment:
Since there's a `return` above, just drop this `else` and reduce the indentation level...